Mathlib/Algebra/Group/EvenFunction.lean

```diff
@@ -0,0 +1,135 @@
+Authors: David Loeffler
+-/
+import Mathlib.Algebra.Group.Action.Pi
+import Mathlib.Algebra.Module.Defs
+
+/-!
+# Even and odd functions
+
+We define even functions `α → β` assuming `α` has a negation, and odd functions assuming both `α`
+and `β` have negation.
+
+These definitions are `Function.Even` and `Function.Odd`; and they are `protected`, to avoid
+conflicting with the root-level definitions `Even` and `Odd` (which, for functions, mean that the
+function takes even resp. odd _values_, a wholly different concept).
+-/
+
+namespace Function
+
+variable {α β : Type*} [Neg α]
+
+/-- A function `f` is _even_ if it satisfies `f (-x) = f x` for all `x`. -/
+protected def Even (f : α → β) : Prop := ∀ a, f (-a) = f a
+
+/-- A function `f` is _odd_ if it satisfies `f (-x) = -f x` for all `x`. -/
+protected def Odd [Neg β] (f : α → β) : Prop := ∀ a, f (-a) = -(f a)
+
+/-- Any constant function is even. -/
+lemma Even.const (b : β) : Function.Even (fun _ : α ↦ b) := fun _ ↦ rfl
+
+/-- The zero function is even. -/
+lemma Even.zero [Zero β] : Function.Even (fun (_ : α) ↦ (0 : β)) := Even.const 0
+
+/-- The zero function is odd. -/
+lemma Odd.zero [NegZeroClass β] : Function.Odd (fun (_ : α) ↦ (0 : β)) := fun _ ↦ neg_zero.symm
+
+section composition
+
+variable {γ : Type*}
+
+/-- If `f` is arbitrary and `g` is even, then `f ∘ g` is even. -/
+lemma Even.left_comp {g : α → β} (hg : g.Even) (f : β → γ) : (f ∘ g).Even :=
+  (congr_arg f <| hg ·)
+
+/-- If `f` is even and `g` is odd, then `f ∘ g` is even. -/
+lemma Even.comp_odd [Neg β] {f : β → γ} (hf : f.Even) {g : α → β} (hg : g.Odd) :
+    (f ∘ g).Even := by
+  intro a
+  simp only [comp_apply, hg a, hf _]
+
+/-- If `f` and `g` are odd, then `f ∘ g` is odd. -/
+lemma Odd.comp_odd [Neg β] [Neg γ] {f : β → γ} (hf : f.Odd) {g : α → β} (hg : g.Odd) :
+    (f ∘ g).Odd := by
+  intro a
+  simp only [comp_apply, hg a, hf _]
+
+end composition
+
+lemma Even.add [Add β] {f g : α → β} (hf : f.Even) (hg : g.Even) : (f + g).Even := by
+  intro a
+  simp only [hf a, hg a, Pi.add_apply]
+
+lemma Odd.add [SubtractionCommMonoid β] {f g : α → β} (hf : f.Odd) (hg : g.Odd) : (f + g).Odd := by
+  intro a
+  simp only [hf a, hg a, Pi.add_apply, neg_add]
+
+section smul
+
+variable {γ : Type*} {f : α → β} {g : α → γ}
+
+lemma Even.smul_even [SMul β γ] (hf : f.Even) (hg : g.Even) : (f • g).Even := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a]
+
+lemma Even.smul_odd [Monoid β] [AddGroup γ] [DistribMulAction β γ] (hf : f.Even) (hg : g.Odd) :
+    (f • g).Odd := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, smul_neg]
+
+lemma Odd.smul_even [Ring β] [AddCommGroup γ] [Module β γ] (hf : f.Odd) (hg : g.Even) :
+    (f • g).Odd := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, neg_smul]
+
+lemma Odd.smul_odd [Ring β] [AddCommGroup γ] [Module β γ] (hf : f.Odd) (hg : g.Odd) :
+    (f • g).Even := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, smul_neg, neg_smul, neg_neg]
```

Comment:
Are you saying that "function • function" should be `smul'`, and "constant • function" should be `smul`? This  convention doesn't seem to be very universal. The scheme I followed, of calling the two `smul` and `const_smul`, matches up with [HasSum.const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/InfiniteSum/Module.html#HasSum.const_smul), [deriv_const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/Deriv/Mul.html#deriv_const_smul), [ConcaveOn.smul_convexOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Mul.html#ConcaveOn.smul_convexOn) and its variants, etc. (There is also [Measurable.const_smul'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/MeasureTheory/Group/Arithmetic.html#Measurable.const_smul') where the first argument is very definitely a constant...)

In the context of `Pi.smul` vs `Pi.smul'`, I thought the difference between `smul'` and `smul` was about the dependence of the _types_, not the values, on the variable.

Reply:
> Are you saying that "function • function" should be `smul'`, and "constant • function" should be `smul`?

Yes

> matches up with [HasSum.const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/InfiniteSum/Module.html#HasSum.const_smul), [deriv_const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/Deriv/Mul.html#deriv_const_smul),

These are so named because they are of the form `c • f x` instead of `c x • f x`

> [ConcaveOn.smul_convexOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Mul.html#ConcaveOn.smul_convexOn) and its variants, etc

This indeed backs up your claim that we are not consistent. Let's get @urkud to tie break?


> I thought the difference between `smul'` and `smul` was about the dependence of the types, not the values, on the variable.

I don't know what you mean here. The difference is whether the action is by scalars or by functions with the same domain.

Reply:
> > matches up with [HasSum.const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/InfiniteSum/Module.html#HasSum.const_smul), [deriv_const_smul](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/Deriv/Mul.html#deriv_const_smul),
> 
> These are so named because they are of the form `c • f x` instead of `c x • f x`

... which is exactly the convention I have followed in the present PR (the lemmas immediately below the lines you highlighted).

Reply:
I don't agree; there is no explicit function application in either the lines this thread is attached to or the ones below.

Reply:
> > I thought the difference between smul' and smul was about the dependence of the types, not the values, on the variable.
> 
> I don't know what you mean here. The difference is whether the action is by scalars or by functions with the same domain.

I mean that `Pi.smul'` allows `c x • f x` where `c` is a dependent function type, i.e. the type of `c x` depends on `x`. But of course a constant value has to have a constant type, so both differences are true and it is a matter of interpretation which is "the" difference.

Regarding "tie breaking": I don't think we have a tie to break. Searching the docs for `smul'` finds many hits, but hardly any of them fit the pattern of using `smul'` for the `c x • f x` usage and `smul` only for `c • f x`. Rather, `xxx_smul'` is used for any kind of slight variation on an existing `xxx_smul` lemma with no pattern at all as to how; whereas we have quite a lot of `const_smul` lemmas.

Reply:
> I mean that `Pi.smul'` allows `c x • f x` where `c` is a dependent function type

I think the confusion here is that I'm talking about the operation that `Pi.smul'` defines, while you're talking about its implementation. I claim that the `smul'` in `Pi.smul'` means "this operation shall be called `smul' elsewhere" and not "this is the name we use for the implementation details of this instance".

I suggest Yury's input since I believe he defined this instance in the first place.

Reply:
>  I claim that the smul' in Pi.smul' means "this operation shall be called `smul' elsewhere" 

And I claim that this attempt to prescribe a convention has not been followed in the subsequent evolution of the library, while the alternative convention of `smul` vs `const_smul` has a bit more traction. Actually, can you point to *any* example in Mathlib where this usage distinction has been observed?

Reply:
More precisely there are 94 declarations whose name contains the string `smul'`. Of these, 29 involve `nsmul'`, `zsmul'`, or `const_smul'` so are definitely not following the alleged convention; and I sampled a selection of the rest and they didn't seem to have anything to do with function types. I have not found a single example, other than `Pi.smul'` itself, where the usage distinction you are advocating is obeyed. Meanwhile, there are 153 declarations involving "const_smul".

Reply:
Yury has confirmed (via Zulip DM's) that the naming of `Pi.smul'` was an arbitrary one-off decision, not intended to set out a convention for the rest of mathlib to follow. So I am going to regard this discussion as resolved.